clash-prelude/src/Clash/Sized/Vector.hs

```diff
@@ -317,25 +324,44 @@ instance KnownNat n => Applicative (Vec n) where
 "zipWith$map" forall f xs ys. zipWith (\g a -> g a) (map f xs) ys = zipWith f xs ys
   #-}
 
-instance (KnownNat n, 1 <= n) => F.Foldable (Vec n) where
-  fold      = leToPlus @1 @n $ fold mappend
-  foldMap f = leToPlus @1 @n $ fold mappend . map f
+instance KnownNat n => F.Foldable (Vec n) where
+  fold Nil      = mempty
+  fold z@Cons{} = fold mappend z
+  foldMap _ Nil      = mempty
+  foldMap f z@Cons{} = fold mappend (map f z)
   foldr     = foldr
   foldl     = foldl
-  foldr1 f  = leToPlus @1 @n $ foldr1 f
-  foldl1 f  = leToPlus @1 @n $ foldl1 f
+  foldr1 _ Nil      = clashCompileError "foldr1: empty Vec"
+  foldr1 f z@Cons{} = foldr1 f z
+  foldl1 _ Nil      = clashCompileError "foldl1: empty Vec"
+  foldl1 f z@Cons{} = foldl1 f z
   toList    = toList
+  null Nil  = True
   null _    = False
   length    = length
-  maximum   = leToPlus @1 @n $ fold (\x y -> if x >= y then x else y)
-  minimum   = leToPlus @1 @n $ fold (\x y -> if x <= y then x else y)
-  sum       = leToPlus @1 @n $ fold (+)
-  product   = leToPlus @1 @n $ fold (*)
+  maximum Nil      = clashCompileError "maximum: empty Vec"
+  maximum z@Cons{} = fold (\x y -> if x >= y then x else y) z
+  minimum Nil      = clashCompileError "minimum: empty Vec"
+  minimum z@Cons{} = fold (\x y -> if x <= y then x else y) z
+  sum Nil      = 0
+  sum z@Cons{} = fold (+) z
+  product Nil      = 1
+  product z@Cons{} = fold (*) z
```

Comment:
So multiplication on `UFixed 0 n` and `SFixed 0 n` / `SFixed 1 n` is a bit weird to begin with, but it does work. However, 1 is not the identity element of multiplication for those types:
```haskell
>>> :set -XNegativeLiterals
>>> -0.5 :: SFixed 0 8
-0.5
>>> as = -0.5 :> -0.5 :> Nil :: Vec 2 (SFixed 0 8)
>>> product as
0.25
>>> 1 * product as
0.12109375
```
This might for instance be a problem in
```haskell
>>> bs = Nil :: Vec 0 (SFixed 0 8)
>>> product (product as :> product bs :> Nil)
0.12109375
```

The question is, do we care? We seem to be violating the laws of multiplication.

Reply:
I guess the question is then more whether `fromInteger` for `SFixed` and `UFixed` should saturate or throw an exception when they go out of bounds.

Reply:
I'm a bit worried about the hidden character of this change. Before, if someone wrote code that say split a vector up so it would end with a `Vec 0` somewhere, it would not type-check. Now, it will type-check, but somehow the outcome is halved. That's less apparent than someone trying to stuff a 1 in `SFixed 0 8` themselves.

Reply:
Would your suggestion then be to have `product` return `clashCompileError` for the `Nil` case instead then?

Reply:
I'm more inclined to change `fromInteger` to use `clashCompileError` for out-of-bounds values, but I'm not saying I have the best answer, just that I think the question is important.

So this is not exactly going to help for this case, but I found it interesting what GHC does for `Data.Int` (I poured a Clash sauce over it to show how it works out for us):
```haskell
module FromInteger where

import Clash.Prelude
import Data.Int

topEntity :: Int8
topEntity = 128
```
gives
```
Loaded package environment from /home/peter/src/clash/clash-compiler/.ghc.environment.x86_64-linux-9.0.2
GHC: Setting up GHC took: 0.219s

FromInteger.hs:7:13: warning: [-Woverflowed-literals]
    Literal 128 is out of the Int8 range -128..127
    If you are trying to write a large negative literal, use NegativeLiterals
  |
7 | topEntity = 128
  |             ^^^

FromInteger.hs:7:13: warning: [-Woverflowed-literals]
    Literal 128 is out of the Int8 range -128..127
    If you are trying to write a large negative literal, use NegativeLiterals
  |
7 | topEntity = 128
  |             ^^^
GHC: Compiling and loading modules took: 0.100s
Clash: Parsing and compiling primitives took 0.181s
GHC+Clash: Loading modules cumulatively took 0.559s
Clash: Compiling FromInteger.topEntity
Clash: Normalization took 0.000s
Clash: Netlist generation took 0.000s
Clash: Compiling FromInteger.topEntity took 0.006s
Clash: Total compilation took 0.566s
```